File: leetcode/46. Permutations.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> res;
        helper(res, nums, 0);
        return res;
    }
private:
    void helper(vector<vector<int>>& res, vector<int>& curr, int m) {
        if(m == curr.size()){
            res.push_back(curr);
            return;
        }
        
        for(int i = m; i < curr.size(); ++i){
            int tmp = curr[m];
            curr[m] = curr[i];
            curr[i] = tmp;
            helper(res, curr, m + 1);
            curr[i] = curr[m];
            curr[m] = tmp;
        }
    }
};

File: leetcode/46. Permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        res = []
        def helper(start):
            if start == len(nums):
                res.append(nums[:])
                return
            
            for i in range(start, len(nums)):
                nums[i], nums[start] = nums[start], nums[i]
                helper(start+1)
                nums[i], nums[start] = nums[start], nums[i]
        
        helper(0)
        return res
